python/dicom_archive.py

```diff
@@ -0,0 +1,325 @@
+
+parser.add_argument(
+    'target',
+    help='The target directory for the DICOM archive')
+
+args = parser.parse_args()
+
+# Typed arguments
+
+profile:   str | None = args.profile
+source:    str        = args.source
+target:    str        = args.target
+verbose:   bool       = args.verbose
+today:     bool       = args.today
+year:      bool       = args.year
+overwrite: bool       = args.overwrite
+db_insert: bool       = args.db_insert
+db_update: bool       = args.db_update
+
+# Check arguments
+
+if db_insert and db_update:
+    print_error_exit(
+        'Arguments \'--db-insert\' and \'--db-update\' must not be set both at the same time.',
+        lib.exitcode.INVALID_ARG,
+    )
+
+if (db_insert or db_update) and not profile:
+    print_error_exit(
+        'Argument \'--profile\' must be set when a \'--db-*\' argument is set.',
+        lib.exitcode.INVALID_ARG,
+    )
+
+if not os.path.isdir(source) or not os.access(source, os.R_OK):
+    print_error_exit(
+        'Argument \'--source\' must be a readable directory path.',
+        lib.exitcode.INVALID_ARG,
+    )
+
+if not os.path.isdir(target) or not os.access(target, os.W_OK):
+    print_error_exit(
+        'Argument \'--target\' must be a writable directory path.',
+        lib.exitcode.INVALID_ARG,
+    )
+
+# Connect to database (if needed)
+
+db = None
+if profile is not None:
+    db = Database(load_config_file(profile).mysql, False)
+    db.connect()
+
+# Check paths
+
+while source.endswith('/'):
+    source = source[:-1]
+
+base_name = os.path.basename(source)
+
+tar_path     = f'{target}/{base_name}.tar'
+zip_path     = f'{target}/{base_name}.tar.gz'
+summary_path = f'{target}/{base_name}.meta'
+log_path     = f'{target}/{base_name}.log'
+
+check_create_file(tar_path)
+check_create_file(zip_path)
+check_create_file(summary_path)
+check_create_file(log_path)
+
+print('Extracting DICOM information (may take a long time)')
+
+summary = lib.dicom.summary_make.make(source, verbose)
+
+if db is not None:
+    print('Checking database presence')
+
+    archive = lib.dicom.dicom_database.get_archive_with_study_uid(db, summary.info.study_uid)
+
+    if db_insert and archive is not None:
+        print_error_exit(
+            (
+                f'Study \'{summary.info.study_uid}\' is already inserted in the database\n'
+                'Previous archiving log:\n'
+                f'{archive[1]}'
+            ),
+            lib.exitcode.INSERT_FAILURE,
+        )
+
+    if db_update and archive is None:
+        print_error_exit(
+            f'No study \'{summary.info.study_uid}\' found in the database',
+            lib.exitcode.UPDATE_FAILURE,
+        )
+else:
+    # Placeholder for type checker
+    archive = None
+
+print('Copying into DICOM tar')
+
+with tarfile.open(tar_path, 'w') as tar:
+    for file in os.listdir(source):
+        tar.add(source + '/' + file)
+
+print('Calculating DICOM tar MD5 sum')
+
+tarball_md5_sum = lib.dicom.text.make_hash(tar_path, True)
```

Comment:
You should use the function `compute_md5_hash` in lib.utilities to compute the hash.

Reply:
I tried to do that and encountered two problems:
- The behaviour of `dicomTar.pl` is to not only write the hash, but also the filename (`"filename    hash"`). `make_hash` emulates this behaviour but not `compute_md5_hash` (although I would not be against writing just the hash).
- `compute_md5_hash` also works on non-existing files, while `make_hash` does not. This makes typing (rightfully) unhappy because I need a `str`, not a `str | None` (`compute_md5_hash` docstring is incorrect with regards to its real return type).
- which makes typing (rightfully) unhappy